File: sdk/typescript/rooch-sdk/src/session/session.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { Args } from '../bcs/index.js'
import { RoochClient } from '../client/index.js'
import { Ed25519Keypair } from '../keypairs/index.js'
import { Transaction } from '../transactions/index.js'
import { BitcoinAddress, RoochAddress } from '../address/index.js'
import { Authenticator, PublicKey, SignatureScheme, Signer } from '../crypto/index.js'

import { CreateSessionArgs } from './types.js'
import { fromHEX } from '../utils/index.js'
import { Bytes } from '../types/index.js'

const DEFAULT_MAX_INACTIVE_INTERVAL = 1200 // second
const REQUIRED_SCOPE = '0x3::session_key::remove_session_key_entry'

type InnerCreateSessionArgs = {
  client: RoochClient
  signer: Signer
} & CreateSessionArgs

type InnerBuildSessionArgs = {
  addr: BitcoinAddress
} & CreateSessionArgs

export class Session extends Signer {
  protected readonly appName: string
  protected readonly appUrl: string
  protected readonly scopes: string[]
  protected readonly keypair: Ed25519Keypair
  protected readonly maxInactiveInterval: number
  protected readonly bitcoinAddress: BitcoinAddress
  protected readonly roochAddress: RoochAddress
  protected lastActiveTime: number

  protected constructor(
    appName: string,
    appUrl: string,
    scopes: string[],
    roochAddress: RoochAddress,
    bitcoinAddress: BitcoinAddress,
    keypair?: Ed25519Keypair,
    maxInactiveInterval?: number,
    localCreateSessionTime?: number,
    lastActiveTime?: number,
  ) {
    super()
    this.appName = appName
    this.appUrl = appUrl
    this.scopes = scopes
    this.roochAddress = roochAddress
    this.bitcoinAddress = bitcoinAddress
    this.keypair = keypair ?? Ed25519Keypair.generate()
    this.maxInactiveInterval = maxInactiveInterval ?? DEFAULT_MAX_INACTIVE_INTERVAL
    this.localCreateSessionTime = localCreateSessionTime ?? Date.now()
    this.lastActiveTime = lastActiveTime || this.localCreateSessionTime
  }

  protected readonly localCreateSessionTime: number

  public static async CREATE(input: InnerCreateSessionArgs): Promise<Session> {
    return this.formatArgs(input, input.signer.getBitcoinAddress()).build(
      input.client,
      input.signer,
    )
  }

  public static Build(input: InnerBuildSessionArgs): string {
    return this.formatArgs(input, input.addr).toJSON()
  }
  static formatArgs(input: CreateSessionArgs, addr: BitcoinAddress): Session {
    const parsedScopes = input.scopes.map((scope) => {
      if (typeof scope !== 'string') {
        return `${scope.address}::${scope.module}::${scope.function}`
      }
      if (scope.split('::').length !== 3) throw Error('invalid scope')
      return scope
    })

    const allOx3 = '0x3::*::*'

    if (!parsedScopes.find((item) => item === allOx3 || item === REQUIRED_SCOPE)) {
      parsedScopes.push(REQUIRED_SCOPE)
    }

    return new Session(
      input.appName,
      input.appUrl,
      parsedScopes,
      addr.genRoochAddress(),
      addr,
      input.keypair,
      input.maxInactiveInterval,
    )
  }

  static fromJson(jsonObj: any) {
    const {
      appName,
      appUrl,
      scopes,
      secretKey,
      maxInactiveInterval,
      bitcoinAddress,
      roochAddress,
      localCreateSessionTime,
      lastActiveTime,
    } = jsonObj

    return new Session(
      appName,
      appUrl,
      scopes,
      new RoochAddress(roochAddress),
      new BitcoinAddress(bitcoinAddress),
      Ed25519Keypair.fromSecretKey(secretKey),
      maxInactiveInterval,
      localCreateSessionTime,
      lastActiveTime,
    )
  }

  sign(input: Bytes): Promise<Bytes> {
    // if (this.lastActiveTime + this.maxInactiveInterval < Date.now() / 1000) {
    //   throw Error('Session is Expired')
    // }
    this.lastActiveTime = Date.now()
    return this.keypair.sign(input)
  }

  signTransaction(input: Transaction): Promise<Authenticator> {
    return Authenticator.rooch(input.hashData(), this)
  }

  getRoochAddress(): RoochAddress {
    return this.roochAddress
  }

  getBitcoinAddress(): BitcoinAddress {
    return this.bitcoinAddress
  }

  getKeyScheme(): SignatureScheme {
    return this.keypair.getKeyScheme()
  }

  getPublicKey(): PublicKey<RoochAddress> {
    return this.keypair.getPublicKey()
  }

  getCreateTime(): number {
    return this.localCreateSessionTime
  }

  getAuthKey(): string {
    return this.keypair.getRoochAddress().toHexAddress()
  }

  protected async build(client: RoochClient, signer: Signer) {
    const [addrs, mods, fns] = this.scopes
      .map((scope) => {
        return scope.split('::')
      })
      .reduce(
        (acc: Array<Array<string>>, val: Array<string>) => {
          acc[0].push(val[0])
          acc[1].push(val[1])
          acc[2].push(val[2])
          return acc
        },
        [[], [], []],
      )

    const tx = new Transaction()
    tx.callFunction({
      target: '0x3::session_key::create_session_key_with_multi_scope_entry',
      args: [
        Args.string(this.appName),
        Args.string(this.appUrl),
        Args.vec('u8', Array.from(fromHEX(this.getAuthKey()))),
        Args.vec('address', addrs),
        Args.vec('string', mods),
        Args.vec('string', fns),
        Args.u64(BigInt(this.maxInactiveInterval)),
      ],
      info: `Welcome to ${this.appName}\nYou will authorize session:\n${
        'Scope:\n' + this.scopes.join('\n') + '\nTimeOut:' + this.maxInactiveInterval.toString()
      }`,
    })

    const result = await client.signAndExecuteTransaction({
      transaction: tx,
      signer: signer,
    })

    if (result.execution_info.status.type === 'executed') {
      return this
    } else {
      throw Error(`create session failed ${result.execution_info.status}`)
    }
  }

  toJSON(): any {
    return {
      appName: this.appName,
      appUrl: this.appUrl,
      scopes: this.scopes,
      secretKey: this.keypair.getSecretKey(),
      maxInactiveInterval: this.maxInactiveInterval,
      bitcoinAddress: this.bitcoinAddress.toStr(),
      roochAddress: this.roochAddress.toStr(),
      localCreateSessionTime: this.localCreateSessionTime,
      lastActiveTime: this.lastActiveTime,
    }
  }
}

File: sdk/golang/rooch-sdk/session/session.go
package session

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/address"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/client"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/crypto"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/crypto/ed25519"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/transactions"
	"github.com/rooch-network/rooch/sdk/golang/rooch-sdk/types"
)

const (
	DefaultMaxInactiveInterval = 1200 // seconds
	RequiredScope             = "0x3::session_key::remove_session_key_entry"
)

type CreateSessionArgs struct {
	AppName            string
	AppUrl             string
	Scopes             []string
	Keypair            *ed25519.Ed25519Keypair
	MaxInactiveInterval int64
}

type Session struct {
	appName             string
	appUrl              string
	scopes              []string
	keypair             *ed25519.Ed25519Keypair
	maxInactiveInterval int64
	bitcoinAddress      *address.BitcoinAddress
	roochAddress        *address.RoochAddress
	localCreateSessionTime int64
	lastActiveTime      int64
}

func CREATE(client *client.RoochClient, signer crypto.Signer, args CreateSessionArgs) (*Session, error) {
	parsedScopes := make([]string, 0)
	for _, scope := range args.Scopes {
		if !strings.Contains(scope, "::") {
			return nil, fmt.Errorf("invalid scope format")
		}
		parsedScopes = append(parsedScopes, scope)
	}

	allOx3 := "0x3::*::*"
	hasRequiredScope := false
	for _, scope := range parsedScopes {
		if scope == allOx3 || scope == RequiredScope {
			hasRequiredScope = true
			break
		}
	}
	if !hasRequiredScope {
		parsedScopes = append(parsedScopes, RequiredScope)
	}

	maxInactiveInterval := args.MaxInactiveInterval
	if maxInactiveInterval == 0 {
		maxInactiveInterval = DefaultMaxInactiveInterval
	}

	keypair := args.Keypair
	if keypair == nil {
		var err error
		keypair, err = ed25519.GenerateKeypair()
		if err != nil {
			return nil, err
		}
	}

	session := &Session{
		appName:             args.AppName,
		appUrl:              args.AppUrl,
		scopes:              parsedScopes,
		keypair:             keypair,
		maxInactiveInterval: maxInactiveInterval,
		bitcoinAddress:      signer.GetBitcoinAddress(),
		roochAddress:        signer.GetRoochAddress(),
		localCreateSessionTime: time.Now().UnixMilli(),
		lastActiveTime:      time.Now().UnixMilli(),
	}

	return session.build(client, signer)
}

func FromJson(jsonData []byte) (*Session, error) {
	var jsonObj map[string]interface{}
	if err := json.Unmarshal(jsonData, &jsonObj); err != nil {
		return nil, err
	}

	keypair, err := ed25519.FromSecretKey([]byte(jsonObj["secretKey"].(string)))
	if err != nil {
		return nil, err
	}

	bitcoinAddr, err := address.NewBitcoinAddress(jsonObj["bitcoinAddress"].(string))
	if err != nil {
		return nil, err
	}

	roochAddr, err := address.NewRoochAddress(jsonObj["roochAddress"].(string))
	if err != nil {
		return nil, err
	}

	scopes := make([]string, 0)
	for _, scope := range jsonObj["scopes"].([]interface{}) {
		scopes = append(scopes, scope.(string))
	}

	return &Session{
		appName:             jsonObj["appName"].(string),
		appUrl:              jsonObj["appUrl"].(string),
		scopes:              scopes,
		keypair:             keypair,
		maxInactiveInterval: int64(jsonObj["maxInactiveInterval"].(float64)),
		bitcoinAddress:      bitcoinAddr,
		roochAddress:        roochAddr,
		localCreateSessionTime: int64(jsonObj["localCreateSessionTime"].(float64)),
		lastActiveTime:      int64(jsonObj["lastActiveTime"].(float64)),
	}, nil
}

func (s *Session) Sign(input []byte) ([]byte, error) {
	s.lastActiveTime = time.Now().UnixMilli()
	return s.keypair.Sign(input)
}

func (s *Session) SignTransaction(tx *transactions.Transaction) (*crypto.Authenticator, error) {
	hashData, err := tx.HashData()
	if err != nil {
		return nil, err
	}
	return crypto.NewRoochAuthenticator(hashData, s)
}

func (s *Session) GetRoochAddress() *address.RoochAddress {
	return s.roochAddress
}

func (s *Session) GetBitcoinAddress() *address.BitcoinAddress {
	return s.bitcoinAddress
}

func (s *Session) GetKeyScheme() crypto.SignatureScheme {
	return s.keypair.GetKeyScheme()
}

func (s *Session) GetPublicKey() crypto.PublicKey {
	return s.keypair.GetPublicKey()
}

func (s *Session) GetCreateTime() int64 {
	return s.localCreateSessionTime
}

func (s *Session) GetAuthKey() string {
	return s.keypair.GetRoochAddress().ToHexAddress()
}

func (s *Session) build(client *client.RoochClient, signer crypto.Signer) (*Session, error) {
	addrs := make([]string, 0)
	mods := make([]string, 0)
	fns := make([]string, 0)

	for _, scope := range s.scopes {
		parts := strings.Split(scope, "::")
		if len(parts) != 3 {
			return nil, fmt.Errorf("invalid scope format")
		}
		addrs = append(addrs, parts[0])
		mods = append(mods, parts[1])
		fns = append(fns, parts[2])
	}

	authKey := s.GetAuthKey()
	authKeyBytes, err := types.FromHex(authKey)
	if err != nil {
		return nil, err
	}

	tx := transactions.NewTransaction()
	info := fmt.Sprintf("Welcome to %s\nYou will authorize session:\nScope:\n%s\nTimeOut:%d",
		s.appName, strings.Join(s.scopes, "\n"), s.maxInactiveInterval)

	tx.CallFunction(transactions.FunctionCall{
		Target: "0x3::session_key::create_session_key_with_multi_scope_entry",
		Args: []interface{}{
			s.appName,
			s.appUrl,
			authKeyBytes,
			addrs,
			mods,
			fns,
			s.maxInactiveInterval,
		},
		Info: info,
	})

	result, err := client.SignAndExecuteTransaction(tx, signer)
	if err != nil {
		return nil, err
	}

	if result.ExecutionInfo.Status.Type != "executed" {
		return nil, fmt.Errorf("create session failed %v", result.ExecutionInfo.Status)
	}

	return s, nil
}

func (s *Session) ToJSON() ([]byte, error) {
	jsonObj := map[string]interface{}{
		"appName":             s.appName,
		"appUrl":              s.appUrl,
		"scopes":              s.scopes,
		"secretKey":           s.keypair.GetSecretKey(),
		"maxInactiveInterval": s.maxInactiveInterval,
		"bitcoinAddress":      s.bitcoinAddress.ToStr(),
		"roochAddress":        s.roochAddress.ToStr(),
		"localCreateSessionTime": s.localCreateSessionTime,
		"lastActiveTime":      s.lastActiveTime,
	}
	return json.Marshal(jsonObj)
} 